### ISSUE TITLE
General question about ENAS

### ISSUE THREAD

**Issue by karandwivedi42** [bkj/ripenet#5]
Hi!

I tried unsuccessfully to replicate the ENAS results before the authors released the official code (using own and your code). Even after running their code and reading the paper, I am not sure about how well it works and had a few general questions.

1. How well do you think the controller works? I am doubtful because the random performance is ~95% - in the paper - so does the controller only offer 1.5% extra accuracy. Did you try the Hyperband, actor-critic etc. and how did it work for you?

2. Were you able to get good accuracy using pytorch code? What do you think you missed? Did you get good accuracy with any other dataset?

Thanks :)

**Comment by bkj** [comment]
Big caveat -- I've only looked at the micro-CNN experiments, and not at all at the RNN experiments.  So everything below may not apply for macro-CNN or RNN experiments.

1) I'm in the middle of doing a bunch of experiments to figure out exactly what's going on in the ENAS code.  But -- I don't think the ENAS controller actually does much.  I did an experiment where i:
```
 - trained 8 random architectures for 10 epochs each
 - picked the one w/ the best validation accuracy
 - trained for 630 epochs, using their `cifar_micro_final.sh` parameters
```
This got to `test_acc=0.9605`, vs. `test_acc=0.9612` for the architecture selected by ENAS.  So not a huge difference -- and definitely stronger than the `test_acc=0.9323` that they report as their random control in the paper.  Given this, I'd expect using hyperband should do about as well as the RL controller, w/ the benefit of being significantly simpler (and more easily parallelizable).

Now, `test_acc=0.9605` is not a bad score -- it's better than a lot of manually designed architectures.  So what's going on?  Not sure exactly, but a couple of clues:
 - 630 epochs is a pretty long time to train a model -- takes ~ 2 days on a TITANX.  Recent [DAWN](https://dawn.cs.stanford.edu/benchmark/CIFAR10/train.html) benchmarks show we can get to ~0.94 accuracy in ~6 minutes.
 - ENAS code uses a bunch of regularization techniques:
    - data augmentation (standard)
    - dropout (lots of networks w/ BN don't use dropout)
    - auxiliary heads 
    - droppath (I'd never seen this before)

In particular, it seems like droppath is critical:
![enas_regularization](https://user-images.githubusercontent.com/6086781/39684471-136d4c72-518a-11e8-833e-b2ee9670ea19.png)
This plot shows performance of the model if you use all three of dropout/aux heads/droppath, and then if you toggle each of them off individually.  Dropout and aux heads don't seem to make much of a difference -- but droppath gives a substantial 1% boost.  I'm currently running experiments to see if adding droppath to a standard ResNet boosts the performance.

2) I haven't gotten results using this code that significantly beat my ResNet18 baseline.  I did try using these methods to find a small architecture to "bolt on" to a pretrained network for finetuning in the CUB bird dataset -- this found architectures w/ modestly better performance than baseline.  Haven't fleshed those experiments out all the way though.

Have you looked at the macro-CNN or RNN experiments?  The RNN results in particular are quite strong, so it would be interesting to have more information about the experimental setup there -- eg, are they using additional tricks that might make it harder to compare their results to the papers they've cited?

~ Ben

**Comment by karandwivedi42** [comment]
I have also looked at macro and micro CNN experiments only. Based on my understanding, 

1. This got to test_acc=0.9605, vs. test_acc=0.9612 for the architecture selected by ENAS. I got similar accuracy using a randomly sampled model. Also, keeping dense skip connections instead of searching for connections also does not harm accuracy.

2. I think the state of the art performance comes more from a smart choice of search space. We tried with CIFAR100/SVHN but did not get above SOTA in any. The controller probably learns a few heuristics, but analyzing the policy also did not indicate that the controller has learned to make specific choices.

3. Overall we observed that CIFAR10 is not a good benchmark as it is too easy (like you said above) and above the 90-93% range we may just be overfitting to the dataset biases instead of learning something new, so we are now looking for challenging (but not as big as imagenet) datasets.

**Comment by bkj** [comment]
1) How did you do your random sampling?  Every random model does not get > 0.95 -- in my experiments, maybe about half of them do.

2) Can you share code for CIFAR100/CVHN experiments?  Did you just swap out the dataset in the authors' ENAS code?

3) How did you come up with the "90-93%" range?  I agree it's possible that the NAS research community has "overfit" to CIFAR10 by using it as the primary dataset.  Unfortunately, since few teams have released their code _and_ the experiments are so costly to run, it's hard to get performance information on other datasets. 

~ Ben

**Comment by karandwivedi42** [comment]
1. Uniform random sampling. I guess I was lucky :stuck_out_tongue: 

2. Yes. We just swapped out the dataset. Also, use stepped LR policy instead of cosine.

3. It was just a guess and I would like to know a better answer (what accuracy is reasonable to expect from a dataset?). For example with wide resnets, keeping [width factor](https://github.com/meliketoy/wide-resnet.pytorch/blob/master/networks/wide_resnet.py#L45)=1 also gets to 91% accuracy (By increasing it all the way to 10, we have 10x increase in capacity and just 4% increase in accuracy). The DAWN Benchmark has 94% accuracy as the threshold. I also came across [this](http://karpathy.github.io/2011/04/27/manually-classifying-cifar10/) which initially got me thinking about whether we are overfitting to dataset biases.

**Comment by karandwivedi42** [comment]
I just used the official code. They have an option to use stepped LR. Can you share your email so I can send you the code? 

I figured that for different datasets the t0 parameter is different (time to the first reset) because sometimes it just resets very fast and doesn't allow the model to reach optimal accuracy before resetting. I would have to search for optimal t0 for each model so I just use stepped. (Also stepped LR reduces one more variable and makes it easy to interpret how your model is learning. I didn't see much loss in accuracy.

I don't think it is easy to get 94-95% accuracy without hyperparameter tuning (if you also take into account the work of authors which we use as default). It is very easy to get >90% though.

**Comment by bkj** [comment]
email: bkj.322@gmail.com -- looking forward to the code.

Re performance baselines: 

[This repo](https://github.com/kuangliu/pytorch-cifar) has a variety of architectures, w/ most getting > 0.93, apparently without a ton of hyperparameter tuning.  

[wide-resnet.pytorch](https://github.com/meliketoy/wide-resnet.pytorch) does even better, though not sure how sensitive those models are to hyperparameter choices.

[fastai](https://github.com/fastai/fastai/blob/848405510b9f3c36eff9af1666514babd8cf165e/courses/dl2/cifar10-dawn.ipynb) has an example of training to > 94% accuracy in 30 epochs (very fast).  Getting good results that fast is probably sensitive to hyperparameters, but if you train for 10x longer I'd bet it's not super sensitive.

Also, re your previous comment: I don't personally think of going from 91% to 95% as "just a 4% increase in accuracy" -- I think about it as "reducing the error rate by 45%".
